### PR TITLE
Update tests in top-level of tests folder for new op-math

### PR DESCRIPTION
**Pull request by lillian542**
For the tests with format `tests/*.py` EXCEPT `tests/test_compile.py` and `tests/test_qaoa.py`

Files that were initially failing:

- [x] `tests/test_device.py`
- [x] `tests/test_qnode.py`
- [x] `tests/test_qnode_legacy.py`
- [x] `tests/test_qubit_device.py`
- [x] `tests/test_queuing.py`
- [x] `tests/test_vqe.py`

These were resolved by a combination of test updates, very small changes to the code made here, and separate changes to the code made in other PRs.

### CODE REVIEW

File: tests/test_device.py [hunk 1]
```diff
@@ -477,6 +495,7 @@ def circuit_gauss(mag_alpha, phase_alpha, phi):
             node_gauss(0.015, 0.02, 0.005)
         assert dev_gauss.num_executions == num_evals_gauss
 
+    @pytest.mark.xfail(reason="behaviour seems different with new op_math")
```
Comment (lillian542) [inline tests/test_device.py:498]:
```suggestion
```

being fixed here: #5379

**Comment by lillian542** [comment]
[sc-58411]

File: pennylane/ops/op_math/linear_combination.py [hunk 2]
```diff
@@ -476,6 +476,7 @@ def diagonalizing_gates(self):
         Returns:
             list[.Operator] or None: a list of operators
         """
+        # raise RuntimeError
```
Comment (lillian542) [inline pennylane/ops/op_math/linear_combination.py:479]:
```suggestion
```

File: tests/test_vqe.py [hunk 3]
```diff
@@ -1027,7 +1033,11 @@ def circuit():
         res = qml.specs(circuit)()
 
         assert res["num_observables"] == 1
-        assert res["num_diagonalizing_gates"] == 0
+        assert (
+            res["num_diagonalizing_gates"] == 0
+            if not H.has_diagonalizing_gates
+            else len(H.diagonalizing_gates())
```
Comment (lillian542) [inline tests/test_vqe.py:1039]: Maybe this is overkill and we should just change it to 1 because that's what's expected in the new op_math version?

**Review by astralcai — COMMENTED** [review]
test_vqe.py seems to fail on my local machine

File: pennylane/gradients/hadamard_gradient.py [hunk 4]
```diff
@@ -329,7 +329,8 @@ def _expval_hadamard_grad(tape, argnum, aux_wire):
                     obs_new = [qml.Z(i) for i in m.wires]
 
                 obs_new.append(qml.Y(aux_wire))
-                obs_new = qml.operation.Tensor(*obs_new)
+                obs_type = qml.prod if qml.operation.active_new_opmath() else qml.operation.Tensor
+                obs_new = obs_type(*obs_new)
```
Comment (astralcai) [inline pennylane/gradients/hadamard_gradient.py:333]: Why not just use the `@` operator?

Comment (lillian542) [reply pennylane/gradients/hadamard_gradient.py:333]: I believe `obs_new` is of varying length, so I'm not sure how to do that nicely

**Comment by lillian542** [comment]
I would prefer this not to be merged until we can discuss the change to `qml.spec` that I'm kind of burying by updating the test, and decide what to do about it (i.e. do we need a story to fix, should it be marked xfail instead of updated, etc)

File: pennylane/gradients/hadamard_gradient.py (hunk 4 above)
Comment (Qottmann) [reply pennylane/gradients/hadamard_gradient.py]: Looks good, saw also christina do that somewhere else, neat solution 👍

File: tests/test_queuing.py [hunk 5]
```diff
@@ -211,6 +211,7 @@ def test_append_tensor_ops(self):
         assert q.queue == [tensor_op]
         assert tensor_op.obs == [A, B]
 
+    @pytest.mark.usefixtures("use_legacy_opmath")
     def test_append_tensor_ops_overloaded(self):
```
Comment (Qottmann) [inline tests/test_queuing.py:215]:
```suggestion
    def test_append_tensor_ops_overloaded_legacy_opmath(self):
```

File: tests/test_vqe.py (hunk 3 above)
Comment (Qottmann) [reply tests/test_vqe.py:1039]: yep I agree

**Review by Qottmann — APPROVED** [review]
Nice! Can confirm all tests `tests/*.py` except for `test_qaoa.py` pass (compiler is skipped in my environment)

Approving with the caveat of making a decision what to do about `qml.spec`

File: tests/test_vqe.py (hunk 3 above)
Comment (Qottmann) [reply tests/test_vqe.py:1039]: Is this the only point needing discussion / making a decision about `qml.specs`? I dont know in detail the context of where this function is used but sounds like a reasonable change to me now that we added `diagonalizing_gates` to `LinearCombination`.

I am going to copy-paste the test to highlight the difference to legacy H though 👍

**Review by astralcai — APPROVED** [review]
Looks good 🎸

File: tests/test_vqe.py (hunk 3 above)
Comment (lillian542) [reply tests/test_vqe.py]: Plan from meeting: Will mark the test as xfail (leaving the expected result as "diagonalizing_gates: 0") for now, and make a shortcut story to revisit this

Comment (Qottmann) [reply tests/test_vqe.py]: sounds good! thanks! Feel free to merge afterwards

**Pull request merged**